File: competitions/SNCKPB17/SNSOCIAL.py

from collections import deque

dx = [0, 1, 1, 1, 0, -1, -1, -1]
dy = [-1, -1, 0, 1, 1, 1, 0, -1]


def find_max(graph):
    ret = 0
    for row in graph:
        ret = max(ret, max(row))

    return ret


def bfs(n, m, graph):
    init_max = find_max(graph)

    visited = set()
    q = deque()
    for i in xrange(n):
        for j in xrange(m):
            if graph[i][j] == init_max:
                q.appendleft(((i, j), 0))
                visited.add((i, j))

    out = 0
    while q:
        (cur_i, cur_j), cur_m = q.pop()
        out = max(out, cur_m)

        for x,y in zip(dx, dy):
            neigh_i = cur_i + y
            neigh_j = cur_j + x

            if neigh_i >= 0 and neigh_i < n and neigh_j >= 0 and neigh_j < m:
                if (neigh_i, neigh_j) not in visited:
                    visited.add((neigh_i, neigh_j))
                    q.appendleft(((neigh_i, neigh_j), cur_m + 1))

    return out


def main():
    t = int(raw_input())

    for _ in range(t):
        n, m = [int(x) for x in raw_input().split()]
        graph = []

        for _ in range(n):
            graph.append([int(x) for x in raw_input().split()])

        print bfs(n, m, graph)


main()

File: codechef/SNCKPB17/SNSOCIAL.cpp

#include <iostream>
#include <queue>
#include <algorithm>
using namespace std;


int graph[501][501];

int dx[] = {0, 1, 1, 1, 0, -1, -1, -1};
int dy[] = {-1, -1, 0, 1, 1, 1, 0, -1};


int main() {
    int t;

    cin >> t;

    while (t--) {
        int n, m;
        cin >> n >> m;

        int init_max = 0;
        for (int i = 0; i < n; ++i)
            for (int j = 0; j < m; ++j) {
                cin >> graph[i][j];
                init_max = max(init_max, graph[i][j]);
            }

        queue<pair<pair<int, int>, int>> q;
        for (int i = 0; i < n; ++i)
            for (int j = 0; j < m; ++j)
                if (graph[i][j] == init_max) {
                    q.push(make_pair(make_pair(i, j), 0));
                    graph[i][j] = -1;
                }

        pair<int, int> coords;
        int cur_i, cur_j, dist, neigh_i, neigh_j;
        int out = 0;
        while (q.size() != 0) {
            coords = q.front().first;
            cur_i = coords.first;
            cur_j = coords.second;
            dist = q.front().second;
            q.pop();

            out = max(out, dist);

            for (int i = 0; i < 8; ++i) {
                neigh_i = cur_i + dy[i];
                neigh_j = cur_j + dx[i];

                if (neigh_i >= 0 and neigh_i < n and neigh_j >= 0 and neigh_j < m)
                    if (graph[neigh_i][neigh_j] != -1) {
                        graph[neigh_i][neigh_j] = -1;
                        q.push(make_pair(make_pair(neigh_i, neigh_j), dist + 1));
                    }
            }
        }

        cout << out << endl;
    }

    return 0;
}
